pkgs/build-support/rust/build-rust-package/default.nix

```
@@ -132,34 +129,36 @@ stdenv.mkDerivation ((removeAttrs args [ "depsExtraArgs" "cargoUpdateHook" "carg
     rustc
   ];
 
-  buildInputs = buildInputs
+  buildInputs = (previousAttrs.buildInputs or [ ])
     ++ lib.optionals stdenv.hostPlatform.isDarwin [ libiconv ]
     ++ lib.optionals stdenv.hostPlatform.isMinGW [ windows.pthreads ];
 
-  patches = cargoPatches ++ patches;
+  patches = cargoPatches ++ (previousAttrs.patches or [ ]);
 
   PKG_CONFIG_ALLOW_CROSS =
     if stdenv.buildPlatform != stdenv.hostPlatform then 1 else 0;
 
   postUnpack = ''
     eval "$cargoDepsHook"
 
-    export RUST_LOG=${logLevel}
-  '' + (args.postUnpack or "");
+    export RUST_LOG=${finalAttrs.logLevel or ""}
+  '' + (previousAttrs.postUnpack or "");
 
-  configurePhase = args.configurePhase or ''
+  configurePhase = previousAttrs.configurePhase or ''
     runHook preConfigure
     runHook postConfigure
   '';
 
-  doCheck = args.doCheck or true;
+  doCheck = previousAttrs.doCheck or true;
 
   strictDeps = true;
 
-  passthru = { inherit cargoDeps; } // (args.passthru or {});
+  passthru = { inherit cargoDeps; } // (previousAttrs.passthru or {});
 
   meta = {
     # default to Rust's platforms
     platforms = rustc.meta.platforms;
-  } // meta;
-})
+  } // (previousAttrs.meta or {});
+};
+
+in args: (stdenv.mkDerivation args).overrideAttrs rustOverride
```

Rather than passing the `buildRustPackage` to `mkDerivation` and then removing them with `removeFromBuilderEnv`. You could remove them here and then pass them to `rustOverride`.

```suggestion
in
args:
let
  mkDerivationArgs = (removeAttrs args [ "depsExtraArgs" "cargoUpdateHook" "cargoLock" ];
in
(stdenv.mkDerivation mkDerivationArgs).overrideAttrs (rustOverride args)
```

---

Yeah, accidentally running into infinite recursions also happened to me a few times :smile: